JSTests/stress/array-isarray-error-message.js fails when run with --useJIT=0.

MenloDorian opened the pull request:
#### 5551d9ec0813336b0daeedee986b101162d2d442
<pre>
JSTests/stress/array-isarray-error-message.js fails when run with --useJIT=0.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247480">https://bugs.webkit.org/show_bug.cgi?id=247480</a>
&lt;rdar://problem/101942000&gt;

Reviewed by Yusuke Suzuki.

isArraySlowInline() was checking if a function is Object.prototype.toString by testing for
function-&gt;intrinsic() == ObjectToStringIntrinsic.  However, this only works when the JIT is enabled.
When the JIT is disabled, function-&gt;intrinsic() is always NoIntrinsic.

The fix is to test function against globalObject-&gt;m_objectProtoToStringFunction instead.  However,
globalObject-&gt;m_objectProtoToStringFunction is a LazyProperty.  We don&apos;t want this check to trigger
the initialization of that LazyProperty, nor do we need to.  Simply testing against its pointer value
is sufficient.  The only way the function can be Object.prototype.toString is if that LazyProperty
has already been initialized.  If it&apos;s not initialized yet, then function cannot be
Object.prototype.toString.  To enable this test, we&apos;ll introduce
JSGlobal::objectProtoToStringFunctionConcurrently().

This issue was discovered by running the JSTests/stress/array-isarray-error-message.js test with the
JIT disabled.

* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::isArraySlowInline):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::objectProtoToStringFunctionConcurrently const):

Canonical link: <a href="https://commits.webkit.org/256314@main">https://commits.webkit.org/256314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0115910d74436ff3a339f9e4c532c528a08c58f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104992 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165252 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4697 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33407 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87776 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100865 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101073 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82022 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/30502 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86478 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81776 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28118 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4362 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40881 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84454 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42870 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39328 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19081 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
```diff
@@ -115,7 +115,7 @@ static ALWAYS_INLINE bool isArraySlowInline(JSGlobalObject* globalObject, ProxyO
             ASCIILiteral calleeName = "Array.isArray"_s;
             auto* function = callee ? jsDynamicCast<JSFunction*>(callee) : nullptr;
-            if (function && function->intrinsic() == ObjectToStringIntrinsic)
+            if (function && function == function->globalObject()->objectProtoToStringFunctionConcurrently())
```
`function` can be ObjectPrototype#toString from different realm. So let's check,

```
function == function->globalObject()->objectProtoToStringFunctionConcurrently()
```

MenloDorian replied:
Thanks.  You beat me to it.  I was also making this change to fix this issue.  It's fixed now with a comment.

Constellation reviewed (approved):
r=me with comment.

Commit:
JSTests/stress/array-isarray-error-message.js fails when run with --useJIT=0.
https://bugs.webkit.org/show_bug.cgi?id=247480
<rdar://problem/101942000>

Reviewed by Yusuke Suzuki.

isArraySlowInline() was checking if a function is Object.prototype.toString by testing for
function->intrinsic() == ObjectToStringIntrinsic.  However, this only works when the JIT is enabled.
When the JIT is disabled, function->intrinsic() is always NoIntrinsic.

The fix is to test function against globalObject->m_objectProtoToStringFunction instead.  However,
globalObject->m_objectProtoToStringFunction is a LazyProperty.  We don't want this check to trigger
the initialization of that LazyProperty, nor do we need to.  Simply testing against its pointer value
is sufficient.  The only way the function can be Object.prototype.toString is if that LazyProperty
has already been initialized.  If it's not initialized yet, then function cannot be
Object.prototype.toString.  To enable this test, we'll introduce
JSGlobal::objectProtoToStringFunctionConcurrently().

This issue was discovered by running the JSTests/stress/array-isarray-error-message.js test with the
JIT disabled.

* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::isArraySlowInline):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::objectProtoToStringFunctionConcurrently const):

Canonical link: https://commits.webkit.org/256314@main
